precompiles/Modexp.yul

```diff
@@ -642,6 +643,70 @@ object "ModExp" {
                 bigUIntDivideNLimbsByTwo(resultPtr, add(nLimbs, nLimbs))
             }
 
+            // @notice Computes the big uint modular exponentiation `result[] := base[] ** exponent[] % modulus[]`.
+            // @param n_limbs Amount of limbs that compose each of the big unsigned integer parameters.
+            // @param base_ptr Base pointer to a big unsigned integer representing the `base[]`. It's most significant half must be zeros.
+            // @param exponent_ptr Base pointer to a big unsigned integer representing the `exponent[]`. It's most significant half must be zeros.
+            // @param modulus_ptr Base pointer to a big unsigned integer representing the `modulus[]`. It's most significant half must be zeros.
+            // @param result_ptr Base pointer to a big unsigned integer to store the result[]. Must be initialized to zeros.
+            function big_uint_modular_exponentiation(n_limbs, base_ptr, exponent_ptr, modulus_ptr, result_ptr) {
+                // Algorithm pseudocode:
+                // See: https://en.wikipedia.org/wiki/Modular_exponentiation#Pseudocode
+                // function modular_pow(base, exponent, modulus) is
+                //     if modulus = 1 then
+                //         return 0
+                //     Assert :: (modulus - 1) * (modulus - 1) does not overflow base
+                //     result := 1
+                //     base := base mod modulus
+                //     while exponent > 0 do
+                //         if (exponent mod 2 == 1) then
+                //             result := (result * base) mod modulus
+                //         exponent := exponent >> 1
+                //         base := (base * base) mod modulus
+                //     return result
+                
+                // PSEUDOCODE: `if modulus = 1 then return 0`.
+                // We are using the precondition that `result[] == [0, ..., 0]`.
+                if big_uint_is_not_one(n_limbs, ptr) {
+
+                    // Assert :: (modulus - 1) * (modulus - 1) does not overflow base
+                    // We are certain that this is true because our precondition requires the most significant half of exponent to be zeros.
+
+                    // PSEUDOCODE: `result := 1`
+                    // Again, we are using the precondition that `result[] == 0`
+                    bigUIntInPlaceOrWith1(result_ptr, n_limbs)
+
+                    // PSEUDOCODE: `base := base mod modulus`
+                    // FIXME: Is ok to mutate the base[] we were given? Shall we use a temporal buffer?
+                    big_uint_mod_inplace(n_limbs, base_ptr, modulus_ptr) // FIXME: not yet implemented.
```

Comment:
We shall discuss if it is **possible** and **acceptable** to mutate `base[]`, which was passed as a parameter.

This depends on how will this function will be used.

- This is to prevent us from requiring a temporary buffer. is it possible? to perform mod in place without one?
  - This helper function will need to call _divmod_ which already requires some temporary buffers.
    - Maybe this memory optimization i'm proposing does not make sense.